### ISSUE TITLE
[Performance] Optional backend capabilities and Reference/FLA speed matrix (V100/4080/4090)

### ISSUE THREAD

**Issue by 123123213weqw** [rwkv-rs/hf-adapter#145]
## 优化后端的定位

当前可选后端通过独立的 `rwkv7-kernels` package 接入，只替换 `ops_rwkv7.py` 的 recurrent operator 边界，不替换 `modeling_rwkv7.py`、HF cache 或模型配置。模型仓库和权重不需要为某张显卡重新制作。

当前已经实现并验证的能力：

- CUDA FP16、`K=V=64`、FP32 recurrent state 的 `native-triton-rank1-scan-v1`；
- prefill、cached decode、attention mask、canonical `[B,H,K,V]` state；
- HF `AutoModelForCausalLM`、cache、generation、beam、save/reload；
- 通过环境变量或 Python context manager 在 reference / optimized 间切换；
- 不支持的 dtype、shape 或训练请求在 `auto` 模式回到标准 PyTorch 路径；
- V100、RTX 4080、RTX 4090 上的正式 `lm_eval` 均为 48/48；
- 安装可选后端后，LoRA SFT/DPO/GRPO 仍可正常运行。

固定强制测试方式：

```bash
RWKV7_BACKEND=optimized RWKV7_KERNEL_IMPL=triton <command>
```

测试中核对的实际实现名称为 `native-triton-rank1-scan-v1`，不是 graph compatibility lane。

## 功能矩阵

| 能力 | Reference | Optional optimized |
|---|---:|---:|
| 标准 HF modeling/config/cache | PASS | 复用同一实现 |
| FP16 CUDA recurrent inference | PASS | PASS |
| Prefill | PASS | PASS |
| Cached decode | PASS | PASS |
| Attention mask / padding | PASS | PASS |
| Batch regrouping | PASS | PASS |
| Greedy / beam generation | PASS | PASS |
| Save / reload | PASS | PASS |
| `lm_eval==0.4.9.1` | 48/48 | V100、4080、4090 均 48/48 |
| SFT / DPO / GRPO | PASS | 安装后 PASS |
| 训练 backward | PyTorch autograd | 自动使用 PyTorch autograd |
| BF16 / FP32 / 其他 shape | PASS | `auto` 使用 reference |

## 0.4B FP16：optimized 相对可读 reference

下表使用每张卡各自的 `speed-0.4b-fp16.json`。数字是 `reference median / optimized median`；`12x` 表示 optimized 用时约为 reference 的 1/12。

| Case | V100 | RTX 4080 | RTX 4090 |
|---|---:|---:|---:|
| `generation_prefill_b1_t1` | 1.086x | 1.056x | 1.059x |
| `generation_prefill_b1_t128` | 12.930x | 11.838x | 14.184x |
| `generation_prefill_b1_t512` | 30.601x | 25.510x | 35.821x |
| `generation_prefill_b1_t2048` | 46.160x | 34.607x | 57.168x |
| `generation_prefill_b4_t128` | 26.617x | 8.471x | 9.857x |
| `generation_prefill_b4_t512` | 40.094x | 10.890x | 13.618x |
| `cached_decode_b1` | 1.094x | 1.046x | 1.072x |
| `cached_decode_b4` | 1.360x | 1.042x | 1.068x |

观察：

- T=1 和 cached decode 主要受整模型其余层与逐 token 调度限制，提升约 1.04x–1.36x；
- T=128–2048 prefill 的 recurrent scan 收益明显，已测区间为 8.47x–57.17x；
- 这些数字比较的是同一张卡、同一环境内的两条路线，不用于直接比较 V100/4080/4090 谁更快。

## V100：WKV operator 与 FLA fused recurrent

V100 记录的是 inference-only fused recurrent 对比。比值列为 `FLA median / optimized median`，大于 1 表示 optimized 更快。

| Shape | Reference ms | Optimized ms | FLA fused ms | Optimized / reference | Optimized / FLA |
|---|---:|---:|---:|---:|---:|
| `b1_t1` | 0.4547 | 0.1782 | 0.1997 | 2.552x | 1.121x |
| `b1_t17` | 4.4155 | 0.1720 | 0.1997 | 25.667x | 1.161x |
| `b1_t128` | 32.1720 | 0.2509 | 0.2806 | 128.237x | 1.118x |
| `b4_t1` | 1.3967 | 0.1638 | 0.1976 | 8.525x | 1.206x |
| `b4_t17` | 18.0378 | 0.1679 | 0.1966 | 107.409x | 1.171x |
| `b4_t128` | 134.7860 | 0.2488 | 0.2796 | 541.675x | 1.123x |

在这组六个 V100 operator shape 上，optimized 相对 FLA fused recurrent 为 1.118x–1.206x。

## RTX 4080：WKV operator 三方矩阵

`Optimized / FLA` 为 `FLA median / optimized median`；大于 1 表示 optimized 更快，小于 1 表示 FLA 更快。

| Shape | Reference ms | Optimized ms | FLA ms | Optimized / reference | Optimized / FLA |
|---|---:|---:|---:|---:|---:|
| `wkv_b1_t128` | 7.7236 | 0.1325 | 0.2386 | 58.309x | 1.801x |
| `wkv_b1_t512` | 30.6661 | 0.3378 | 0.2290 | 90.769x | 0.678x |
| `wkv_b1_t2048` | 122.1544 | 1.1806 | 0.3058 | 103.466x | 0.259x |
| `wkv_b4_t128` | 9.9961 | 0.1090 | 0.2177 | 91.684x | 1.997x |
| `wkv_b4_t512` | 39.7285 | 0.3218 | 0.2211 | 123.442x | 0.687x |

## RTX 4080：0.4B whole-model 三方矩阵

| Case | Reference ms | Optimized ms | FLA ms | Optimized / reference | Optimized / FLA |
|---|---:|---:|---:|---:|---:|
| `prefill_no_cache_b1_t128` | 226.417 | 19.580 | 23.519 | 11.563x | 1.201x |
| `prefill_no_cache_b1_t512` | 857.683 | 34.847 | 24.573 | 24.613x | 0.705x |
| `prefill_no_cache_b1_t2048` | 3367.325 | 102.413 | 104.635 | 32.880x | 1.022x |
| `prefill_no_cache_b4_t128` | 296.360 | 36.172 | 24.686 | 8.193x | 0.682x |
| `prefill_no_cache_b4_t512` | 1134.914 | 107.973 | 103.101 | 10.511x | 0.955x |
| `cached_decode_b1` | 1422.031 | 1371.599 | 975.360 | 1.037x | 0.711x |
| `cached_decode_b4` | 1455.691 | 1402.529 | 1012.504 | 1.038x | 0.722x |

## RTX 4090：WKV operator 三方矩阵

`Optimized / FLA` 为 `FLA median / optimized median`；大于 1 表示 optimized 更快，小于 1 表示 FLA 更快。

| Shape | Reference ms | Optimized ms | FLA ms | Optimized / reference | Optimized / FLA |
|---|---:|---:|---:|---:|---:|
| `wkv_b1_t128` | 31.2455 | 0.1945 | 0.6372 | 160.615x | 3.276x |
| `wkv_b1_t512` | 125.4441 | 0.4083 | 0.6294 | 307.228x | 1.541x |
| `wkv_b1_t2048` | 501.9994 | 1.2813 | 0.6634 | 391.775x | 0.518x |
| `wkv_b4_t128` | 39.8563 | 0.1890 | 0.6505 | 210.928x | 3.442x |
| `wkv_b4_t512` | 158.3045 | 0.4054 | 0.6502 | 390.495x | 1.604x |

## RTX 4090：0.4B whole-model 三方矩阵

| Case | Reference ms | Optimized ms | FLA ms | Optimized / reference | Optimized / FLA |
|---|---:|---:|---:|---:|---:|
| `prefill_no_cache_b1_t128` | 892.097 | 63.726 | 59.598 | 13.999x | 0.935x |
| `prefill_no_cache_b1_t512` | 3417.616 | 99.877 | 60.977 | 34.218x | 0.611x |
| `prefill_no_cache_b1_t2048` | 13551.036 | 246.444 | 62.255 | 54.986x | 0.253x |
| `prefill_no_cache_b4_t128` | 1144.710 | 121.939 | 61.217 | 9.388x | 0.502x |
| `prefill_no_cache_b4_t512` | 4422.531 | 329.812 | 62.491 | 13.409x | 0.189x |
| `cached_decode_b1` | 4887.441 | 4612.760 | 2597.167 | 1.060x | 0.563x |
| `cached_decode_b4` | 4945.669 | 4662.267 | 2709.329 | 1.061x | 0.581x |

## Shape 结论

| 硬件 / 层级 | Optimized 更有优势的已测 shape | FLA 更有优势的已测 shape |
|---|---|---|
| V100 WKV operator | B1/B4，T=1/17/128 全部六项 | 本轮未出现 |
| RTX 4080 WKV operator | B1T128、B4T128 | B1T512、B1T2048、B4T512 |
| RTX 4080 whole model | B1T128、B1T2048 | B1T512、B4T128、B4T512、decode B1/B4 |
| RTX 4090 WKV operator | B1T128、B1T512、B4T128、B4T512 | B1T2048 |
| RTX 4090 whole model | — | 本轮全部七项 |

这说明当前 native scan 的强项是把可读逐 token recurrence 压缩成单次 device-side scan，尤其适合中短 operator shape 和 reference prefill 加速；whole-model 最终速度仍受投影、归一化、FFN、数据布局及后端整体融合程度影响。FLA 在部分长序列和 whole-model case 中具有更完整的融合优势。

## `lm_eval` 与训练生态矩阵

| Gate | V100 | RTX 4080 | RTX 4090 |
|---|---:|---:|---:|
| Forced optimized route | PASS | PASS | PASS |
| 正式 `lm_eval` | 48/48 | 48/48 | 48/48 |
| batch 1/8 metric validator | PASS | PASS | PASS |
| HF AutoModel/generation/save-reload/cache | PASS | PASS | PASS |
| Training autograd fallback | PASS | PASS | PASS |
| LoRA SFT/DPO/GRPO smoke | PASS | PASS | PASS |
| Adapter reload | `max_abs=0.0` | `max_abs=0.0` | `max_abs=0.0` |

## 复现信息

- 代码：`cfeb5aeca860ce444ebb3515a20cc22f7e2b090b`；
- 证据提交：`0c5ea30ac6868974ba9836c4a065fa8b2847af68`；
- FLA：`80e494f6c588e091fc8316b612870df29375c5b8`；
- kernel wheel SHA256：`544b08225351d9658a1941c260a70710d47bb677d2976d730cec9e4f34771560`；
- 0.4B FP16；warmup 2，median of 5；decode 64 tokens；model-level CUDA Graph 与 `torch.compile` 关闭。

环境不同，因此跨 GPU 的绝对毫秒数不作为显卡排名：

- V100：PyTorch 2.5.1+cu124；
- RTX 4080：PyTorch 2.11.0+cu130，Triton 3.6.0；
- RTX 4090 三方对比：PyTorch 2.7.1+cu126，Triton 3.3.1。

完整证据：

- [三卡汇总](https://github.com/123123213weqw/hf-adapter/blob/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/TRITON_CFEB5AE_SUMMARY.md)
- [V100 bundle](https://github.com/123123213weqw/hf-adapter/tree/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/v100-cfeb5ae-triton)
- [RTX 4080 bundle](https://github.com/123123213weqw/hf-adapter/tree/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/4080-cfeb5ae-triton)
- [RTX 4090 bundle](https://github.com/123123213weqw/hf-adapter/tree/0c5ea30ac6868974ba9836c4a065fa8b2847af68/results/native-backend/4090-cfeb5ae-triton)

后续可以基于这份 shape 矩阵继续讨论独立 native、FLA 或混合 dispatch 路线；新增结论应继续附相同 shape、环境、代码 SHA 和原始 JSON。
